lib/discourse_auth_cookie.rb

```diff
@@ -0,0 +1,138 @@
+# frozen_string_literal: true
+
+class DiscourseAuthCookie
+  class InvalidCookie < StandardError; end
+
+  class Encryptor
+    def encrypt_and_sign(plain_cookie)
+      encryptor.encrypt_and_sign(plain_cookie)
+    end
+
+    def decrypt_and_verify(cipher_cookie)
+      encryptor.decrypt_and_verify(cipher_cookie)
+    rescue ActiveSupport::MessageVerifier::InvalidSignature
+      raise InvalidCookie.new
+    end
+
+    private
+
+    def encryptor
+      @encryptor ||= begin
+        key = Rails.application.key_generator.generate_key("discourse-auth-cookie", 32)
+        ActiveSupport::MessageEncryptor.new(
+          key,
+          key,
+          cipher: "aes-256-cbc",
+          digest: "SHA256"
+        )
+      end
+    end
+  end
+
+  TOKEN_SIZE ||= 32
+
+  TOKEN_KEY ||= "token"
+  ID_KEY ||= "id"
+  TL_KEY ||= "tl"
+  TIME_KEY ||= "time"
+  VALID_KEY ||= "valid"
+  private_constant *%i[
+    TOKEN_KEY
+    ID_KEY
+    TL_KEY
+    TIME_KEY
+    VALID_KEY
+  ]
+
+  attr_reader *%i[token user_id trust_level timestamp valid_for]
+
+  def self.parse(raw_cookie, secret = Rails.application.secret_key_base)
+    # v0 of the cookie was simply the auth token itself. we need this for
+    # backward compatibility so we don't wipe out existing sessions
+    return new(token: raw_cookie) if raw_cookie.size == TOKEN_SIZE
+
+    data = Encryptor.new.decrypt_and_verify(raw_cookie)
+    # data, sig = raw_cookie.split("|", 2)
+    # validate_signature!(data, sig, secret)
+
+    token = nil
+    user_id = nil
+    trust_level = nil
+    timestamp = nil
+    valid_for = nil
+
+    data.split(",").each do |part|
+      prefix, val = part.split(":", 2)
+      val = val.presence
+      if prefix == TOKEN_KEY
+        token = val
+      elsif prefix == ID_KEY
+        user_id = val
+      elsif prefix == TL_KEY
+        trust_level = val
+      elsif prefix == TIME_KEY
+        timestamp = val
+      elsif prefix == VALID_KEY
+        valid_for = val
+      end
+    end
+
+    new(
+      token: token,
+      user_id: user_id,
+      trust_level: trust_level,
+      timestamp: timestamp,
+      valid_for: valid_for,
+    )
+  end
+
+  def self.validate_signature!(data, sig, secret)
+    data = data.to_s
+    sig = sig.to_s
+    if compute_signature(data, secret) != sig
+      raise InvalidCookie.new
+    end
+  end
+
+  def self.compute_signature(data, secret)
+    OpenSSL::HMAC.hexdigest("sha256", secret, data)
+  end
+
+  def initialize(token:, user_id: nil, trust_level: nil, timestamp: nil, valid_for: nil)
+    @token = token
+    @user_id = user_id.to_i if user_id
+    @trust_level = trust_level.to_i if trust_level
+    @timestamp = timestamp.to_i if timestamp
+    @valid_for = valid_for.to_i if valid_for
```

Comment:
This field is included in the `_t` cookie and is set equal to the `maximum_session_age` site setting at the time the cookie is generated. When parsing the cookie, a check is performed to ensure `timestamp + valid_for >= Time.zone.now`.

We could remove this field and compare the `timestamp` field directly with the site setting, but _I think_ there is a small chance for a db query to be made when looking up a site setting so we'd need to move the failover middleware before request tracker.

The downside of including cookie age in the cookie is the site setting could be changed but existing cookies that were generated before the setting was changed would still be validated using whatever they have in the `valid_for` field (you wouldn't be able to actually login with such cookie but the app would do user-based rate limiting instead of IP-based since it thinks the cookie is valid).

Reply:
I think we keep it in the cookie, it is more robust. The impact of it being wrong is not too high. If people must they can always just log everyone out.